### PR TITLE
Lifted while loop

### PR DESCRIPTION
**Pull request by jheek**
- adds while_loop to lift.py
- Adds mutable_filter to scope_fn.
  This way collections can be made
  immutable in part of the lifted transform.
  (Like the carry variables in the cond_fn of a while_loop)
- Adds while_loop to transforms.py
- Adds multi fns lifted transform support

### CODE REVIEW

File: flax/core/lift.py [hunk 1]
```diff
@@ -720,7 +721,90 @@ def scanned(broadcast_vars, carry, scan_variable_groups, rng_groups, args):
...
+
+  def inner(scope_fn, repack_fn,
+            variable_groups, rng_groups, carry_init):
+    carry_variables, broadcast_variables = variable_groups
+
+    def make_loop_rngs(i):
+      local_rng_groups = []
+      for rng_group, rng_split in zip(rng_groups, rng_splits):
+        if rng_split:
+          rng_group = tree_map_rngs(lambda rng: random.fold_in(rng, -i), rng_group)
```
Comment (levskaya) [inline flax/core/lift.py:779]: why flip the sign?

**Review by levskaya — APPROVED** [review]
Looks great, just a few things that might need deleted?

File: flax/core/lift.py (hunk 1 above)
Comment (jheek) [reply flax/core/lift.py:779]: Otherwise cond_fn and body_fn share rngs.

Comment (levskaya) [reply flax/core/lift.py:773]: sure, I guess I was wondering why have the negative here rather than just below.  it's not important :)

Comment (jheek) [reply flax/core/lift.py:773]: Oh I see you're right. This one is redundant

**Pull request merged**